### ISSUE TITLE
Feature suggestion: Total Variation Denoising

### ISSUE THREAD

**Issue by Hvass-Labs** [tensorflow/tensorflow#6397]
I would like to suggest the following feature be added to TensorFlow and discuss it before making the pull-request. As far as I can tell, the feature does not exist in TensorFlow.

### Motivation

Total Variation Denoising (TVD) is sometimes used as a regularizer in image processing to suppress noise. It is commonly used in Style Transfer implementations. Here is my recent tutorial which uses it:

https://github.com/Hvass-Labs/TensorFlow-Tutorials/blob/master/15_Style_Transfer.ipynb

While researching this tutorial, I saw some bizarre implementations of TVD. In reality it is just a simple array-slicing operation. I thought it would be useful to include it in TensorFlow so it is easy to use for everyone.

### Math Formula

The code below implements the 'anisotropic' version of the formula, because it should be easier to optimize. The formula can be seen here:

https://en.wikipedia.org/wiki/Total_variation_denoising#2D_digital_signals

### Code

This is the code I propose to add to TensorFlow:

    def total_variation(images):
        """
        Calculate the Total Variation for one or more images for use in denoising.

        https://en.wikipedia.org/wiki/Total_variation_denoising

        Args:
            images: `Tensor` with one or more images.
                    The shape is `[batch, height, width, channel]`.

        Returns:
            A scalar `Tensor` representing the value.
        """

        value = tf.reduce_sum(tf.abs(images[:,1:,:,:] - images[:,:-1,:,:])) + \
                tf.reduce_sum(tf.abs(images[:,:,1:,:] - images[:,:,:-1,:]))
    
        return value

Style note: For more complicated expressions like this, I prefer to assign the value before I return it. It makes the code cleaner and is helpful in debugging.

### Discussion

I studied [this TensorFlow module](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/losses/losses.py) but did not fully understand the design-philosophy and how to make this new function fit in.

Before making the pull-request I have some questions and issues to discuss:

1) Would this belong under losses or regularizers in TensorFlow? Where specifically?

2) Should I add a scope-block?

3) I made an implementation that only reduced the sum for `axis=[1, 2, 3]` thus calculating the value on a per-image basis in case there are multiple images in the batch. But I don't know if it would ever be used in that way and it made the implementation more complicated. Perhaps it is best to have the simple version above and extend it if necessary.

**Comment by aselle** [comment]
It seems reasonable to me. @martinwicke, do you have any opinions here, or do you know who might?

**Comment by martinwicke** [comment]
People who may have an informed opinion on this: @shlens, @sguada, @vincentvanhoucke.

To your questions: 

1) This function should probably go into the image module, since it's specific to image data.
2) Yes, please add a name_scope. It'll make the graph more readable in tensorboard.
3) It would be good to have a batch version which returns a vector. I would expect to be used more than the single image version.

**Comment by shlens** [comment]
SGTM. +1 to everything  @martinwicke suggested.

I would suggest adding the function to:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/image_ops_impl.py 

The function should check that its argument `images` is of the appropriate shape using `_Check3DImage` or `_CheckAtLeast3DImage`

**Comment by shlens** [comment]
@Hvass-Labs Would you be able to make a proper pull request for this code? Please include a unit test in:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/image_ops_test.py.

Note that the second argument to `images` seems to have more indices then the dimensions. Not sure if this is a potential bug.

**Comment by Hvass-Labs** [comment]
I have now opened a pull request https://github.com/tensorflow/tensorflow/pull/6662

@shlens I don't understand your remark about a potential bug.

Please have a look at the PR and let's continue the discussion there.
